src/gui/windows/controller_mappings_settings.zig

```diff
@@ -0,0 +1,56 @@
+const std = @import("std");
+
+const main = @import("root");
+const files = main.files;
+const download_controller_mappings = @import("download_controller_mappings.zig");
+const settings = main.settings;
+const Vec2f = main.vec.Vec2f;
+
+const gui = @import("../gui.zig");
+const GuiComponent = gui.GuiComponent;
+const GuiWindow = gui.GuiWindow;
+const Button = @import("../components/Button.zig");
+const CheckBox = @import("../components/CheckBox.zig");
+const Label = @import("../components/Label.zig");
+const VerticalList = @import("../components/VerticalList.zig");
+const HorizontalList = @import("../components/HorizontalList.zig");
+
+pub var window = GuiWindow {
+	.contentSize = Vec2f{256, 256},
+	.closeable = false
+};
+
+const padding: f32 = 8;
+
+fn askCallback(value: bool) void {
+	settings.askToDownloadControllerMappings = !value;
+}
+fn unrecognizedCallback(value: bool) void {
+	settings.downloadControllerMappingsWhenUnrecognized = value;
+}
+fn periodicCallback(value: bool) void {
+	settings.downloadControllerMappings = value;
+}
+fn onSave(_: usize) void {
+	settings.save();
+	gui.closeWindowFromRef(&window);
+	gui.openWindow("download_controller_mappings");
+}
+pub fn onOpen() void {
+	const list = VerticalList.init(.{padding, 16 + padding}, 300, 16);
+	list.add(Label.init(.{0, 0}, 256, "Controller Mapping Settings", .center));
+	list.add(CheckBox.init(.{0, 0}, 256, "Don't ask at startup", !settings.askToDownloadControllerMappings, &askCallback));
+	list.add(CheckBox.init(.{0, 0}, 256, "Download controller mappings periodically", settings.downloadControllerMappings, &periodicCallback));
+	list.add(CheckBox.init(.{0, 0}, 256, "Download controller mappings when unrecognized controller plugged in", settings.downloadControllerMappingsWhenUnrecognized, &unrecognizedCallback));
+	list.add(Button.initText(.{0, 0}, 256, "Close", .{.callback = &onSave}));
```

Comment:
None of the other menus have a separate close button. Why should this one have it?

You can still put all your behavior into the `onClose()` function.

Reply:
I'll see if this works. When I last tried that, opening a window in onClose would fail, for unknown reasons.

Reply:
It still does, which is a bug. I think I'll temporarily keep the close button until that is fixed. I might fix it in another PR.

Reply:
I also can't get the window to open in onClose. I need the button to open the controller mapping download window, when it's needed.

Reply:
I guess another question would be: Does this window even make sense? Is the fact that we download the controller mappings even useful information to the average player?

Reply:
Sorry, your replies didn't show up until I reloaded the window. The controller mapping download window makes sense as when the user has extremely slow download speeds (like me), it might take a few seconds or longer. I'm not entirely sure about the controller mapping download settings window.

Reply:
I guess it could be used to allow the user to supply custom mappings in the case that their controller isn't in the download.

Reply:
The controller mappings need to be downloaded as early as possible, otherwise the controller can't be used at all. It affects startup times. The thread is used to prevent the UI from freezing when downloading.

Reply:
That way there can be an indication that something is actually happening, and the game isn't broken

Reply:
Perhaps we should download controller mappings without asking, in case the user only wants to use a controller with the game.

Reply:
It may make sense to download in the background, then. I think I'll have an indication that it's happening just in case, but it'll be kept out of the way.